File: 总题库/125.ValidPalindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int n = s.length();
        if (n < 2) return true;

        int low = 0, high = n - 1;
        while (low < high) {
            if (!isdigit(s[low]) && !isalpha(s[low])) {
                low++;
                continue;
            }
            if (!isdigit(s[high]) && !isalpha(s[high])) {
                high--;
                continue;
            }
            if (tolower(s[low++]) != tolower(s[high--])) return false;
        }
        return true;
    }
};

File: 总题库/125.ValidPalindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        n = len(s)
        if n < 2:
            return True
        
        s = s.lower()
        low = 0
        high = n - 1
        while low < high:
            if (s[low] < 'a' or s[low] > 'z') and (s[low] > '9' or s[low] < '0'):
                low += 1
                continue
            if (s[high] < 'a' or s[high] > 'z') and (s[high] > '9' or s[high] < '0'):
                high -= 1
                continue
            if s[low] != s[high]:
                print(s[low], s[high])
                return False
            low += 1
            high -= 1
        return True